File: Mobile/lib/core/models/product_model.dart
import 'package:nexora/core/entities/product.dart';

class ProductModel extends Product {
  ProductModel({
    required super.id,
    required super.name,
    required super.brand,
    required super.description,
    required super.price,
    required super.discount,
    required super.stock,
    required super.sold,
    required super.ratingRate,
    required super.ratingCount,
    required super.categoryId,
    required super.categoryName,
    required super.thumbnail,
    required super.images,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    return ProductModel(
      id: json["_id"] ?? "",
      name: json["name"] ?? "Unknown Product",
      brand: json["brand"] ?? "Unknown Brand",
      description: json["description"] ?? "No description available.",
      price: json["price"] ?? 0,
      discount: json["discount"] ?? 0,
      stock: json["stock"] ?? 0,
      sold: json["sold"] ?? 0,
      ratingRate: json["rating"]?["rate"] ?? 0,
      ratingCount: json["rating"]?["count"] ?? 0,
      categoryId: json["category"]?["_id"] ?? "",
      categoryName: json["category"]?["name"] ?? "Unknown Category",
      thumbnail: json["thumbnail"] ?? "",
      images: json["images"] != null ? List<String>.from(json["images"]) : [],
    );
  }
}

File: Backend/models/productModel.js
import mongoose from "mongoose";

const productSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, "Product name is required"],
      trim: true,
      minlength: [3, "Product name must be at least 3 characters long"],
    },
    brand: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Brand",
      required: [true, "Product brand is required"],
    },
    description: {
      type: String,
      required: [true, "Product description is required"],
      trim: true,
      minlength: [4, "Product description must be at least 4 characters long"],
    },
    price: {
      type: Number,
      required: [true, "Product price is required"],
      min: [0, "Price must be a positive number"],
    },
    discount: {
      type: Number,
      min: [0, "Discount must be a positive number"],
      max: [100, "Discount can't exceed 100%"],
      default: 0,
    },
    stock: {
      type: Number,
      required: [true, "Product stock is required"],
      min: [0, "Stock must be a positive number"],
    },
    sold: { type: Number, default: 0 },
    rating: {
      rate: {
        type: Number,
        min: [0, "Rating must be a positive number"],
        max: [5, "Rating can't exceed 5"],
        default: 0,
      },
      count: {
        type: Number,
        default: 0,
        min: [0, "Rating count must be a positive number"],
      },
    },
    category: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Category",
      required: [true, "Product category is required"],
    },
    thumbnail: {
      type: String,
      required: [true, "Product thumbnail is required"],
    },
    images: {
      type: [String],
      required: [true, "Product images are required"],
    },
  },
  {
    timestamps: true,
  },
);

const Product = mongoose.model("Product", productSchema);

export default Product;
